File: pa2/Lex.c
/****************************************************************************************
*  Nicholas Pappas
*  nhpappas@ucsc.edu
*  CMPS 101 w/ Tantalo
*  Assignment: pa1
*****************************************************************************************/
//-----------------------------------------------------------------------------
// Lex.c
// Illustrates file input-output commands and text processing using the
// string functions strtok and strcat.
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 400

int main(int argc, char * argv[]){

   int i, n=0, count=0;
   FILE *in, *out;
   char line[MAX_LEN];

   // check command line for correct number of arguments
   if( argc != 3 ){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if( out==NULL ){
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }

   List A = newList();
   /* read each line of input file, & count num lines*/
   while( fgets(line, MAX_LEN, in) != NULL)  {
      count++;
   }
   //close and open the infile
   fclose(in);
   in = fopen(argv[1], "r");
   
      char word[count-1][MAX_LEN];
   /* read lines into word[]*/
    n=0;
   while( fgets(line, MAX_LEN, in) != NULL)  {
      //printf("%s",line);
      strcpy(word[n],line);
      n++;
   }
   //printf("File input done\n");
   append(A, 0);
   /*for(i=0; i<count; i++){
        printf("%s\n",word[i]);
    }*/
   for (i= 1; i<count; i++){
	   moveFront(A);
	   while(index(A) != -1){
          // printf("%s",word[i]);
		   if( strcmp(word[i], word[get(A)]) < 0 ){
		   //s1.compareTo(s2)<0 is true if and only if s1 comes before s2
		      insertBefore(A, i);
		      break;
		   }else{
			  moveNext(A);
		   }
		   
		   if(index(A) == -1){
              append(A, i);
              break;
		   }
	   }
   }
   moveFront(A);
   while( index(A) != -1 ){
	   fprintf(out, "%s", word[get(A)]);
	   moveNext(A);
   }
   freeList(&A);
   /* close files */
   fclose(in);
   fclose(out);

   return(0);
}

File: pa1/Lex.java
/*	Nicholas Pappas
	nhpappas@ucsc.edu
	ID: 1558554
	CMPS 101 w/ Tantalo MWF 10:40-11:45AM
	Assignment: pa1
*/
// Lex.java
// Illustrates file IO and List implementation
import java.io.*;
import java.util.Scanner;
class Lex{

 public static void main(String[] args) throws IOException{ 
   Scanner in = null;
   PrintWriter out = null;
   String line = null;
   int j, n, lineNumber = 0;
 // check number of command line arguments is at least 2
   if(args.length < 2){
      System.out.println("Usage: FileCopy <input file> <output file>");
      System.exit(1);
    }
 // open files
   in = new Scanner(new File(args[0]));
   out = new PrintWriter(new FileWriter(args[1]));
 // read lines from in
   while( in.hasNextLine() ){ 
      lineNumber++;	
	  }			// Count num lines in file so we can							
   String[] word = new String[lineNumber]; // allocate an array of that size.
   List A = new List();
   lineNumber = 0;				// Reuse lineNumber as an index for line
   in.close();
   in = null;
   in = new Scanner(new File(args[0]));
   while( in.hasNextLine() ){  // trim leading and trailing spaces, then add one
      line = in.nextLine().trim() + " "; // trailing space so split works on blank lines
	  word[lineNumber] = line;		//load lines from file into the array we allocated	   
	  lineNumber++;
   }
   A.prepend(0);				//Put an initial element in the list to compareTo
   for(j=1; j < lineNumber; j++){ //Step each element through the list to find its place
	  n = 0; 
	  A.moveBack();
	  while(n == 0){
	     if(word[j].compareTo(word[A.index()]) < 0){	//j comes before and isn't one of these cases
            if(A.length() == 1 || A.index() == 0){ 
			A.insertBefore(j); 
			n = -1;
			}else{ A.movePrev(); //move the cursor. Check again incase we can insert.
			   if(word[j].compareTo(word[A.index()]) > 0){ 
			   A.insertAfter(j); 
			   n = -1; }
			}
	     }else if(word[j].compareTo(word[A.index()]) > 0){	//j comes after and isn't one of these cases
		    if(A.length() == 1 || A.index() == A.length()-1){ 
			A.insertAfter(j);
			n = -1; 
		    }else{ A.moveNext(); //move the cursor. Check again incase we can insert.
			   if(word[j].compareTo(word[A.index()]) < 0){ 
			   A.insertBefore(j); 
			   n = -1; }
			}
		 } //The while loop will run this logic untill the element is inserted.
	  } 
   }
   
   for(A.moveFront(); A.index()>=0; A.moveNext()){ //Print out the array in sorted order
      out.print(word[A.get()]+"\n");
   }
 // close files
   in.close();
   out.close();
   }
}